sympy/polys/matrices/domainmatrix.py

```diff
@@ -119,6 +122,19 @@ def __rmul__(A, B):
         else:
             return NotImplemented
 
+    def __smul__(self, other):
+        if not isinstance(other, DomainScalar):
+            other = DomainScalar(other, self.domain)
+
+        self, other = self.unify(other)
+        mat = self.rep.to_list()
+        rows, cols = self.shape
+        ele = other.element
+        for r in range(rows):
+            for c in range(cols):
+                mat[r][c] = mat[r][c]*ele
+        return DomainMatrix.from_list_sympy(rows, cols, mat)
```

Comment:
Converting to a list is inefficient. The `to_list` and `from_list_sympy` methods convert from the domain to Expr and back. The whole point of `DomainMatrix` is that we do all the calculations with the more efficient domain elements and not with Expr so this is backwards.

Scalar multiplication should be delegated to the `rep` so that `DDM` and `SDM` can have different implementations. (The sparse implementation should not use a dense algorithm like this)

Reply:
There is one confusion, when the element has no specified domain, then what domain should we assign to it?
Like for example, `S(3)/2` what domain should we assign to it, so that after unifying we will get `QQ`, as the domain in:
```
In [1]: from sympy.polys.matrices import DomainMatrix
   ...: dM = DomainMatrix.from_Matrix(Matrix([[1, 2], [3, 4]]))
   ...: a=(S(3)/2)
   ...: dM.__smul__(a)
Out[1]: DomainMatrix([[3/2, 3], [9/2, 6]], (2, 2), QQ)
```

Reply:
@oscarbenjamin, kindly review the latest commit, it adds scalar multiplication for all three classes using rep as it is implemented for other methods.
Kindly share your opinion.
Thanks.